Sources/NIO/BaseSocketChannel.swift

```diff
@@ -205,19 +205,84 @@ private struct SocketChannelLifecycleManager {
 class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, ChannelCore {
     typealias SelectableType = SocketType.SelectableType
 
-    // MARK: Stored Properties
-    // Visible to access from EventLoop directly
+    struct AddressCache {
+        // deliberately lets because they must always be updated together (so forcing `init` is useful).
+        let local: Optional<SocketAddress>
+        let remote: Optional<SocketAddress>
+
+        init(local: SocketAddress?, remote: SocketAddress?) {
+            self.local = local
+            self.remote = remote
+        }
+    }
+
+    // MARK: - Stored Properties
+    // MARK: Constants & atomics (accessible everywhere)
     public let parent: Channel?
     internal let socket: SocketType
     private let closePromise: EventLoopPromise<Void>
     internal let selectableEventLoop: SelectableEventLoop
-    private let addressesCached: AtomicBox<Box<(local:SocketAddress?, remote:SocketAddress?)>> = AtomicBox(value: Box((local: nil, remote: nil)))
-    private let bufferAllocatorCached: AtomicBox<Box<ByteBufferAllocator>>
+    private let _offEventLoopLock = Lock()
     private let isActiveAtomic: NIOAtomic<Bool> = .makeAtomic(value: false)
-    private var _pipeline: ChannelPipeline! = nil // this is really a constant (set in .init) but needs `self` to be constructed and therefore a `var`. Do not change as this needs to accessed from arbitrary threads
     // just a thread-safe way of having something to print about the socket from any thread
     internal let socketDescription: String
 
+    // MARK: Variables, on EventLoop thread only
+    var readPending = false
+    var pendingConnect: Optional<EventLoopPromise<Void>>
+    var recvAllocator: RecvByteBufferAllocator
+    var maxMessagesPerRead: UInt = 4
+    private var inFlushNow: Bool = false // Guard against re-entrance of flushNow() method.
+    private var autoRead: Bool = true
+
+    // MARK: Varibles that are really constants
+    private var _pipeline: ChannelPipeline! = nil // this is really a constant (set in .init) but needs `self` to be constructed and therefore a `var`. Do not change as this needs to accessed from arbitrary threads
+
+    // MARK: Special variables, please read comments.
+    // For reads guarded by _offEventLoopLock/EL thread; writes guarded by _offEventLoopLock (and on EL thread only).
```

Comment:
Can we reword this slightly to make it clearer? What we want to say, as I understand it, is:

> "Reads require holding either the `_offEventLoopLock` or EL thread, writes require holding both.